### PR TITLE
SI-7015 Removes redundant aconst_null; pop; aconst_null creation

### PR DESCRIPTION
**Pull request by JamesIry**
In an effort to adapt methods and field accesses of type Null to
other types, we were always emitting

```
aconst_null
pop
aconst_null
```

The problem is we were doing that even when the JVM was in a position
to know it had null value, e.g. when the user had written a null
constant. This commit fixes that and includes a test to show that the
resulting byte code still works even without repeating ourselves and/or
repeating ourselves.

This commit also makes the scala.runtim.Null$ constructor private. It
was a sealed abstract class which prevented subclassing in Scala, but
it didn't prevent subclassing in Java. A private constructor takes care
of that hole so now the only value of type Null$ should be null.

Along the way I found some other questionable things in adapt and I've
added TODO's and issue https://issues.scala-lang.org/browse/SI-7159 to
track.

### CODE REVIEW

**Comment by JamesIry** [comment]
So apparently the in GenICode#adapt is dealing with expressions of type Null, which in turn erase to type scala.runtime.Null$, which of course the JVM thinks of as being a type incompatible with anything but java.lang.Object. GenICode#adapt can't distinguish between "no, really, it's a null constant" vs "it's the result of an expression of type Null and therefore we need to do some chicanery to make the JVM happy." It should be easy to fix the case where the expression is the constant null.

**Comment by adriaanm** [comment]
>  the case where the expression is the constant null.

... or evaluates to it?

**Comment by JamesIry** [comment]
Yeah, I think I only really ever need to adapt method calls and field/array element refs. Block expressions, if/else, etc.  of type Null should then make the JVM happy because the JVM does have an internal null type that it uses for bytecode verification and if those other expressions are adapted then that's the type JVM will see.

**Comment by JamesIry** [comment]
review @gkossakowski

File: src/compiler/scala/tools/nsc/backend/icode/GenICode.scala [hunk 1]
```diff
@@ -997,13 +1005,40 @@ abstract class GenICode extends SubComponent  {
...
+     * If we have a method call, field load, or array element load of type Null then
+     * we need to convince the JVM that we have a null value because in Scala
+     * land Null is a subtype of all ref types, but in JVM land scala.runtime.Null$
+     * is not. Note we don't have to adapt loads of locals because the JVM type
+     * system for locals does have a null type which it tracks internally. As
+     * long as we adapt these other things, the JVM will know that a Scala local of
+     * type Null is holding a null.
+     */
+    private def adaptNullRef(from: TypeKind, to: TypeKind, ctx: Context, pos: Position) {
+      log(s"GenICode#adaptNullRef($from, $to, $ctx, $pos)")
```
Comment (gkossakowski) [inline src/compiler/scala/tools/nsc/backend/icode/GenICode.scala:1019]: This very much looks like a candidate for a macro to me but we'll have to wait until macros stabilize.

File: src/compiler/scala/tools/nsc/backend/icode/GenICode.scala [hunk 2]
```diff
@@ -997,13 +1005,40 @@ abstract class GenICode extends SubComponent  {
...
+      
+      // Don't need to adapt null to unit because we'll just drop it anyway. Don't
+      // need to adapt to Object or AnyRef because the JVM is happy with 
+      // upcasting Null to them.
+      // We do have to adapt from NullReference to NullReference because we could be storing
+      // this value into a local of type Null and we want the JVM to see that it's
+      // a null value so we don't have to also adapt local loads.
+      if (from == NullReference && to != UNIT && to != ObjectReference && to != AnyRefReference) {
+        assert(to.isReferenceType, "Attempt to adapt a null to a non reference type")
```
Comment (gkossakowski) [inline src/compiler/scala/tools/nsc/backend/icode/GenICode.scala]: Shouldn't assert message include the type? I know we log it above but this log must be explicitly enabled by user.

File: test/files/run/t7015.scala [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  def r(x: AnyRef) = x
+  def s(x: Any) = x
+}
+
+abstract class A {
+	def f(x: String): String
+}
+
+class B extends A {
+	val nullField = null
```
Comment (gkossakowski) [inline test/files/run/t7015.scala:41]: Shouldn't we declare type explicitly as `Null` or you really want it to be inferred?

**Comment by gkossakowski** [comment]
Thanks for putting comments that highly illuminating. Once two my comments are addressed this is good to go.

File: test/files/run/t7015.scala (hunk 3 above)
Comment (JamesIry) [inline test/files/run/t7015.scala:41]: Inferred or not doesn't affect the test, so probably clearer to explicitly have the type.

**Comment by gkossakowski** [comment]
LGTM!

**Pull request merged**